streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1144,4 +1154,74 @@ public void cleanUp() {
+     *
+     * Note: Each invocation of this method issues a call to the Kafka brokers. Thus its advisable to limit the frequency
+     * of invocation to once every few seconds.
+     *
+     * @return map of store names to another map of partition to {@link LagInfo}s
+     */
+    public Map<String, Map<Integer, LagInfo>> allLocalStorePartitionLags() {
+        final long latestSentinel = -2L;
+        final Map<String, Map<Integer, LagInfo>> localStorePartitionLags = new TreeMap<>();
+
+        final Collection<TopicPartition> allPartitions = new LinkedList<>();
+        final Map<TopicPartition, Long> changelogPositions = new HashMap<>();
+
+        // Obtain the current positions, of all the active-restoring and standby tasks
+        for (final StreamThread streamThread : threads) {
+            for (final StandbyTask standbyTask : streamThread.allStandbyTasks()) {
+                allPartitions.addAll(standbyTask.changelogPartitions());
+                // Note that not all changelog partitions, will have positions; since some may not have started
+                changelogPositions.putAll(standbyTask.getChangelogPositions());
+            }
+
+            final Set<TaskId> restoringTaskIds = streamThread.restoringTaskIds();
+            for (final StreamTask activeTask : streamThread.allStreamsTasks()) {
+                final Collection<TopicPartition> taskChangelogPartitions = activeTask.changelogPartitions();
+                allPartitions.addAll(taskChangelogPartitions);
+
+                final boolean isRestoring = restoringTaskIds.contains(activeTask.id());
+                final Map<TopicPartition, Long> restoredOffsets = activeTask.restoredOffsets();
+                for (final TopicPartition topicPartition : taskChangelogPartitions) {
+                    if (isRestoring && restoredOffsets.containsKey(topicPartition)) {
+                        changelogPositions.put(topicPartition, restoredOffsets.get(topicPartition));
+                    } else {
+                        changelogPositions.put(topicPartition, latestSentinel);
+                    }
+                }
+            }
+        }
+
+        log.debug("Current changelog positions: {}", changelogPositions);
+        final Map<TopicPartition, ListOffsetsResultInfo> allEndOffsets;
+        try {
+            allEndOffsets = adminClient.listOffsets(
+                allPartitions.stream()
+                    .collect(Collectors.toMap(Function.identity(), tp -> OffsetSpec.latest()))
+            ).all().get();
+        } catch (final RuntimeException | InterruptedException | ExecutionException e) {
+            throw new StreamsException("Unable to obtain end offsets from kafka", e);
+        }
+        log.debug("Current end offsets :{}", allEndOffsets);
+        for (final Map.Entry<TopicPartition, ListOffsetsResultInfo> entry : allEndOffsets.entrySet()) {
+            // Avoiding an extra admin API lookup by computing lags for not-yet-started restorations
```

Comment:
@vvcephei @brary I actually have a concern with this overestimation.. Let's take a typical streams deployment with num_standby_replicas =1 ... During the time period, where the lag is fetched when a standby changelog partition is assigned but has not yet processed any events, we will report a really high value that may not lead to IQs being failed (since the store is deemed laggy).. Given a caller that periodically fetches the positions every few seconds would have a non-zero past value already, wonder if it makes sense to not report this partition instead.. i.e we only report lags on store partitions which have begun consumption..  At least for ksqlDB, I am thinking servers can track easily when lag was reported for a given store partition, given server and decide on using that value based on how recently it was updated..  But the values we receive won't be 'jumpy'. End of the day, I am also theorizing how this will behave.. So this is more about which way we want to bias. 


Across the cluster, this time period will not in sync with each other and can happen differently. So not sure if this will in-fact provide correct relative order..

Reply:
>but it does correctly state that the second instance is ahead of the first in freshness.

Fair. Your argument is if you have not begun work, then you would be behind another instance that has begun work.. 

>reasonable to not report a lag at all for not-yet-started stores, but then you have to handle the special case in client-side code.

the client code is going to have to use some mechanism to "fence" really bad lag.. i.e full restorations.. My concern was that if we hit this race condition, you would suddenly see a jump to "oh this replica needs to fully restore" and then back to the actual lag..

>One other thought is that offset values can still be jumpy, due to compaction.

I did think about this a bit.. I think what we have here would work fairly well in the common case, where the active and standby are apart by lag within the `uncompacted` portion of the changelog topic.. 

> So there's nothing to say that there are any records at all between offset 5 and offset 9999.
Once beyond that, since we just rely on the latest lag value, this is okay.. if there are no records, the restoration will be faster than expected.. but as a relative metric, it's still good.. 

In all, I am okay sticking with this behavior for now.. We can revisit based on real world experience.. and really build a solid solid implementation in the next month or so..